core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java

```diff
@@ -467,18 +477,21 @@ public void run() {
                         }
                         boolean success = false;
                         try {
-                            // connect to the node, see if we manage to do it, if not, bail
-                            if (!nodeFoundByAddress) {
-                                logger.trace("[{}] connecting (light) to {}", sendPingsHandler.id(), finalNodeToSend);
-                                transportService.connectToNodeAndHandshake(finalNodeToSend, timeout.getMillis());
+                            Transport.Connection connection;
+                            if (nodeFoundByAddress && transportService.nodeConnected(finalNodeToSend)) {
```

Comment:
How can this returns true? This code is only executed if further up `!transportService.nodeConnected(nodeToSend)` holds (or we created a temporary disco node which has a fresh id and is never connected).

I also wonder if we should use a temporary disco-node when `nodeFoundByAddress == true` and `transportService.nodeConnected(nodeToSend) == false`.

Last, regarding the existing comment on line 454
> // if we find on the disco nodes a matching node by address, we are going to restore the connection
>  // anyhow down the line if its not connected...

Now that we are not reestablishing the full connection to a node anymore that's in the CS, I wonder what this will do when the node gets disconnected but not removed from the cluster state. When pinging completes (with a temporary connection) and the node stays in the cluster state that is published by the master, then the master will not reestablish the connection in ClusterService (as the node has not been removed on the CS and connectToNodes is only called on the newly added nodes).

Let's talk about this